### PR TITLE
Add `addScriptAnchor` to control where in the page the autogen bundle is placed (#1578).

### PR DESCRIPTION
**Pull request by Fizzixnerd**
Addresses #1578.

When creating a `Widget`, `addScriptAnchor` can be used to anchor the autogen bundle to a particular location in the page.  If omitted, the default behavior is to put the script at the end of the bundle as before.

Usage:

```haskell
myWidget = do
  addScript $ StaticR js_before_bundle_js
  addScriptAnchor
  addScript $ StaticR js_after_bundle_js
```

- [x] Bumped the version number
- [x] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [x] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddocks for new, public APIs

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [x] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

### CODE REVIEW

**Comment by Fizzixnerd** [comment]
So the CI failed on one of the Linux GHCs.  It seems like it couldn't find fpcomplete.com?

File: yesod-core/Yesod/Core/Class/Yesod.hs [hunk 1]
```diff
@@ -541,18 +546,24 @@ widgetToPageContent w = HandlerFor $ \hd -> do
                    $ encodeUtf8 $ renderJavascriptUrl render s
                 return $ renderLoc x
 
-    -- modernizr should be at the end of the <head> http://www.modernizr.com/docs/#installing
-    -- the asynchronous loader means your page doesn't have to wait for all the js to load
-    let (mcomplete, asyncScripts) = asyncHelper render scripts jscript jsLoc
+    let scripts'' = catMaybes $ (\case
```
Comment (snoyberg) [inline yesod-core/Yesod/Core/Class/Yesod.hs]: This looks like a usage of `mapMaybe`. It's probably even more elegant as a list comprehension `[s | SOAScript s <- scripts]`.

**Comment by Fizzixnerd** [comment]
I have added tests to the suite to demonstrate it working.

The mechanism is basically this: in the code below in `regularScriptLoad`, it used to make a script tag for each script, and then at the end put the autogen bundle.  Now it places the bundle at the location of any SOAAnchors it finds.
```haskell
    let scripts'' = [s | SOAScript s <- scripts]
        -- modernizr should be at the end of the <head> http://www.modernizr.com/docs/#installing
        -- the asynchronous loader means your page doesn't have to wait for all the js to load
        (mcomplete, asyncScripts) = asyncHelper render scripts'' jscript jsLoc
        regularScriptLoad = [hamlet|
            $newline never
            $forall s <- scripts
                $case s
                    $of SOAScript s
                        ^{mkScriptTag s}
                    $of SOAAnchor
                        $maybe j <- jscript
                            $maybe s <- jsLoc
                                <script src="#{s}" *{jsAttributes master}>
                            $nothing
                                <script>^{jelper j}
        |]
```

The reason this is okay is because just above we have:
```haskell
      scripts = runUniqueList (scripts' <> toUnique SOAAnchor)
```
which makes sure that:

1. There is at most one `SOAAnchor` so the bundle is not included multiple times.
2. There is at least one `SOAAnchor` at the end of the bundle, in case `addScriptAnchor` was never used.  This preserves the old behavior for old code.

The only thing I'm just now wondering about is if there would be any difference for async scripts.

**Comment by snoyberg** [comment]
I may be dense here, but is there a reason not to go in the direction I mentioned, of generating `script` tags and adding them as HTML to the body?

**Comment by Fizzixnerd** [comment]
One issue is that there is no instance of `Blaze.ToMarkup` for JSON `Value`s.  For instance, if I wanted do something like:
```haskell
data MyObj = MyObj
  { s :: String,
    n :: Int
  } deriving (Eq, Show)

instance ToJSON MyObj where
  toJSON MyObj {..} = object
    [ "s" .= s
    , "n" .= n
    ]

embeddedScript :: Handler Html
embeddedScript = defaultLayout $ do
  let myObj = MyObj "string and stuff" 4
  [whamlet|
          <script>
            myObj = #{toJSON myObj};
          |]
```
then I get a type error, while

```haskell
normalScript :: Handler Html
normalScript = defaultLayout $ do
  let myObj = MyObj "string and stuff" 4
  toWidget [julius|myObj = #{toJSON myObj};|]
```
works just fine.  It is, of course, possible to serialize this value manually, but more complicated examples would be more annoying to deal with (such as an array, which is something I actually have to pass in my application).

**Comment by jezen** [comment]
Closing due to inactivity.

**Pull request closed without merging**